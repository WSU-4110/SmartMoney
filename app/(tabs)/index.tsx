// Import statements
import React, { FC, useEffect, useState } from 'react';
import {
  View,
  Text,
  SafeAreaView,
  ScrollView,
  StyleSheet,
  Dimensions,
  TouchableOpacity,
  RefreshControl, // Import RefreshControl
} from 'react-native';
import { Colors } from '@/constants/Colors';
import { useColorScheme } from '@/hooks/useColorScheme';
import { PieChart } from 'react-native-gifted-charts';
import AsyncStorage from '@react-native-async-storage/async-storage';
import api from '@/api/apiClient';

// Get device screen dimensions
const { width } = Dimensions.get('window');

// Interface definitions
interface Account {
  id: string;
  name: string;
  balance: number;
}

interface CreditCard {
  id: string;
  name: string;
  balance: number;
  available: number;
}

interface Institution {
  id: string;
  name: string;
  accounts: Account[];
  creditCards: CreditCard[];
}

const Page: FC = () => {
  const colorScheme = useColorScheme();
  const currentColors = Colors[colorScheme ?? 'light'];

  const [institutions, setInstitutions] = useState<Institution[]>([]);
  const [totalBalance, setTotalBalance] = useState(0);

  // State for refreshing
  const [refreshing, setRefreshing] = useState(false);

  const fetchData = async () => {
    try {
      const userId = await AsyncStorage.getItem('userId');
      if (!userId) return;

      // Fetch institutions
      const institutionResponse = await api.post('/user/userinfo', { id: userId });
      const institutionsData = institutionResponse.data.items || [];
      let calculatedBalance = 0;
      console.log('Institutions:', institutionsData);

      const formattedInstitutions: Institution[] = institutionsData.map((institution: any) => {
        const accounts = institution.accounts.map((account: any) => ({
          id: account.id,
          name: account.name,
          balance: account.balances[0]?.current || 0,
        }));

        const creditCards =
          institution.creditCards?.map((card: any) => ({
            id: card.id,
            name: card.name,
            balance: card.balances?.current || 0,
            available: card.balances?.available || 0,
          })) || [];

        // Calculate total balance
        calculatedBalance += accounts.reduce((sum: number, acc: Account) => sum + acc.balance, 0);

        return {
          id: institution.institution.id,
          name: institution.institution.name,
          accounts,
          creditCards,
        };
      });

      setInstitutions(formattedInstitutions);
      setTotalBalance(calculatedBalance);
    } catch (error) {
      console.error('Error fetching data:', error);
    }
  };

  useEffect(() => {
    fetchData();
  }, []);

  // Pull-to-refresh handler
  const onRefresh = async () => {
    setRefreshing(true);
    await fetchData();
    setRefreshing(false);
  };

  return (
    <SafeAreaView style={[styles.container, { backgroundColor: currentColors.background }]}>
      <ScrollView
        showsVerticalScrollIndicator={false}
        contentContainerStyle={styles.scrollContent}
        refreshControl={
          <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
        }
      >
        {/* Interactive Graph Section */}
        <View style={styles.graphContainer}>
          <Text style={[styles.sectionTitle, { color: currentColors.text }]}>Overview</Text>
          <PieChart
            data={institutions.map((inst, index) => ({
              value: inst.accounts.reduce((sum, acc) => sum + acc.balance, 0),
              color: [currentColors.accent, currentColors.primary, currentColors.secondary][index % 3],
              label: inst.name,
            }))}
            donut
            radius={150}
            innerRadius={100}
            innerCircleColor={currentColors.background}
            centerLabelComponent={() => (
              <View style={styles.centerLabel}>
                <Text style={[styles.centerLabelAmount, { color: currentColors.text }]}>
                  ${totalBalance.toFixed(2)}
                </Text>
                <Text style={[styles.centerLabelText, { color: currentColors.text }]}>
                  Total Balance
                </Text>
              </View>
            )}
          />
        </View>

        {/* Accounts Summary by Institution */}
        <View style={styles.accountsContainer}>
          <Text style={[styles.sectionTitle, { color: currentColors.text }]}>Accounts</Text>
          {institutions.length > 0 ? (
            institutions.map((inst) => (
              <View key={inst.id} style={styles.institutionContainer}>
                <Text style={[styles.institutionName, { color: currentColors.accent }]}>
                  {inst.name}
                </Text>
                {inst.accounts.map((account) => (
                  <TouchableOpacity key={account.id} style={styles.accountItem}>
                    <Text style={[styles.accountName, { color: currentColors.text }]}>
                      {account.name}
                    </Text>
                    <Text style={[styles.accountBalance, { color: currentColors.primary }]}>
                      ${account.balance.toFixed(2)}
                    </Text>
                  </TouchableOpacity>
                ))}
              </View>
            ))
          ) : (
            <Text style={[styles.noDataText, { color: currentColors.text }]}>
              No accounts available.
            </Text>
          )}
        </View>

        {/* Credit Cards Summary */}
        <View style={styles.creditCardsContainer}>
          <Text style={[styles.sectionTitle, { color: currentColors.text }]}>Credit Cards</Text>
          {institutions.flatMap((inst) => inst.creditCards).length > 0 ? (
            institutions.flatMap((inst) => inst.creditCards).map((card) => (
              <TouchableOpacity key={card.id} style={styles.cardItem}>
                <View>
                  <Text style={[styles.cardName, { color: currentColors.text }]}>{card.name}</Text>
                  <Text style={[styles.cardAvailable, { color: currentColors.secondary }]}>
                    Available: ${card.available.toFixed(2)}
                  </Text>
                </View>
                <Text style={[styles.cardBalance, { color: currentColors.accent }]}>
                  Balance: ${card.balance.toFixed(2)}
                </Text>
              </TouchableOpacity>
            ))
          ) : (
            <Text style={[styles.noDataText, { color: currentColors.text }]}>
              No credit cards available.
            </Text>
          )}
        </View>
      </ScrollView>
    </SafeAreaView>
  );
};

// Styles
const styles = StyleSheet.create({
  container: { flex: 1 },
  scrollContent: { paddingVertical: 20, paddingBottom: 70 },
  graphContainer: { alignItems: 'center', marginVertical: 20 },
  centerLabel: { justifyContent: 'center', alignItems: 'center' },
  centerLabelAmount: { fontSize: 24, fontWeight: '600' },
  centerLabelText: { fontSize: 16, color: '#666' },
  sectionTitle: { fontSize: 22, fontWeight: '600', marginBottom: 10 },
  accountsContainer: { marginHorizontal: 20, marginVertical: 10 },
  institutionContainer: { marginBottom: 20 },
  institutionName: { fontSize: 18, fontWeight: '700', marginBottom: 10 },
  accountItem: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingVertical: 15,
    borderBottomColor: '#ccc',
    borderBottomWidth: StyleSheet.hairlineWidth,
  },
  accountName: { fontSize: 16 },
  accountBalance: { fontSize: 16, fontWeight: '500' },
  creditCardsContainer: { marginHorizontal: 20, marginVertical: 10 },
  cardItem: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingVertical: 15,
    borderBottomColor: '#ccc',
    borderBottomWidth: StyleSheet.hairlineWidth,
  },
  cardName: { fontSize: 16 },
  cardAvailable: { fontSize: 14, color: '#666' },
  cardBalance: { fontSize: 16, fontWeight: '500' },
  noDataText: { fontSize: 16, fontStyle: 'italic', textAlign: 'center', marginTop: 10 },
});

export default Page;
