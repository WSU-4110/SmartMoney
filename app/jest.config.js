module.exports = {
    preset: 'jest-expo',
    setupFiles: ['<rootDir>/jest.setup.js'],
    transformIgnorePatterns: [
      'node_modules/(?!((jest-)?react-native|@react-native(-community)?)|expo(nent)?|@expo(nent)?/.*|@expo-google-fonts/.*|react-navigation|@react-navigation/.*|@unimodules/.*|unimodules|sentry-expo|native-base|react-native-svg)'
    ],
    // Add these lines:
    maxWorkers: 1,  // Run tests serially
    testEnvironment: 'node',
    testTimeout: 30000,  // Increase timeout
  };s