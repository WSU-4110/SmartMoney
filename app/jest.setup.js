// jest.setup.js
import '@testing-library/jest-native/extend-expect';

// Mock AsyncStorage
jest.mock('@react-native-async-storage/async-storage', () => require('./__mocks__/@react-native-async-storage/async-storage'));