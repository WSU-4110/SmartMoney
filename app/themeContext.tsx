import React, { createContext, useContext, useState, ReactNode, useEffect } from 'react';
import { DarkTheme, DefaultTheme } from '@react-navigation/native';
import AsyncStorage from '@react-native-async-storage/async-storage';

interface ThemeContextType {
  theme: typeof DefaultTheme;
  toggleTheme: () => void;
}

const ThemeContext = createContext<ThemeContextType | undefined>(undefined);

interface ThemeProviderProps {
  children: ReactNode;
}

export const ThemeProvider: React.FC<ThemeProviderProps> = ({ children }) => {
  const { theme, toggleTheme } = usePersistedTheme();
  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};

export const useTheme = () => {
  const context = useContext(ThemeContext);
  if (!context) {
    throw new Error('useTheme must be used within a ThemeProvider');
  }
  return context;
};

const THEME_STORAGE_KEY = 'appTheme';

export const usePersistedTheme = () => {
  const [theme, setTheme] = useState(DefaultTheme);

  useEffect(() => {
    const loadTheme = async () => {
      const savedTheme = await AsyncStorage.getItem(THEME_STORAGE_KEY);
      if (savedTheme) {
        setTheme(savedTheme === 'dark' ? DarkTheme : DefaultTheme);
      }
    };
    loadTheme();
  }, []);

  const toggleTheme = async () => {
    const newTheme = theme.dark ? DefaultTheme : DarkTheme;
    setTheme(newTheme);
    await AsyncStorage.setItem(THEME_STORAGE_KEY, newTheme.dark ? 'dark' : 'light');
  };

  return { theme, toggleTheme };
};
