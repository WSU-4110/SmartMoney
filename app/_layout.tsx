import React, { useEffect, useState } from 'react';
import { ThemeProvider as NavigationThemeProvider, DarkTheme, DefaultTheme } from '@react-navigation/native';
import { useFonts } from 'expo-font';
import { Stack } from 'expo-router';
import * as SplashScreen from 'expo-splash-screen';
import { ThemeProvider } from './themeContext'; // Ensure this path is correct

// Prevent the splash screen from auto-hiding before asset loading is complete.
SplashScreen.preventAutoHideAsync();

export default function RootLayout() {
  const [loaded] = useFonts({
    SpaceMono: require('../assets/fonts/SpaceMono-Regular.ttf'),
  });

  const [theme, setTheme] = useState(DefaultTheme); // Initialize theme state
  const toggleTheme = () => {
    setTheme((prevTheme) => (prevTheme.dark ? DefaultTheme : DarkTheme));
  };

  useEffect(() => {
    if (loaded) {
      SplashScreen.hideAsync();
    }
  }, [loaded]);

  // Ensure all hooks are called regardless of loading state
  return (
    <ThemeProvider value={{ theme, toggleTheme }}>
      <NavigationThemeProvider value={theme}>
        <Stack>
          <Stack.Screen name="(tabs)" options={{ headerShown: false }} />
          <Stack.Screen name="+not-found" />
        </Stack>
      </NavigationThemeProvider>
    </ThemeProvider>
  );
}
