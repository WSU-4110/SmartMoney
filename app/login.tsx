import React from 'react';
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  SafeAreaView,
  Image,
} from 'react-native';
import { Colors } from '@/constants/Colors';
import { useColorScheme } from '@/hooks/useColorScheme';
import { FontAwesome } from '@expo/vector-icons';
import { useAuth } from '@/app/auth/AuthProvider';
import { router } from 'expo-router';
import { useTheme } from './themeContext';


export default function LoginScreen(): JSX.Element {
  const { login } = useAuth();
  const { theme, toggleTheme } = useTheme(); 
  const currentColors = Colors[theme.dark ? 'dark' : 'light'];

  const handleLogin = async () => {
    await login();
    router.replace('/(tabs)');
  };

  return (
    <SafeAreaView style={[styles.container, { backgroundColor: currentColors.background }]}>
      <View style={styles.content}>
        <View style={styles.logoContainer}>
          <Image 
            source={require('@/assets/images/SmartMoneySplashwSign.png')}
            style={styles.logo}
            resizeMode="contain"
          />
          <Text style={[styles.appName, { color: currentColors.text }]}>
            SmartMoney
          </Text>
          <Text style={[styles.tagline, { color: currentColors.secondary }]}>
            Your Personal Finance Tracker
          </Text>
        </View>

        <View style={styles.featuresContainer}>
          <View style={styles.featureItem}>
            <FontAwesome name="bar-chart" size={24} color={currentColors.primary} />
            <Text style={[styles.featureText, { color: currentColors.text }]}>
              Track your expenses
            </Text>
          </View>
          <View style={styles.featureItem}>
            <FontAwesome name="bank" size={24} color={currentColors.primary} />
            <Text style={[styles.featureText, { color: currentColors.text }]}>
              Manage multiple accounts
            </Text>
          </View>
          <View style={styles.featureItem}>
            <FontAwesome name="bell" size={24} color={currentColors.primary} />
            <Text style={[styles.featureText, { color: currentColors.text }]}>
              Get important alerts
            </Text>
          </View>
        </View>

        <TouchableOpacity
          style={[styles.loginButton, { backgroundColor: currentColors.primary }]}
          onPress={handleLogin}
        >
          <FontAwesome name="sign-in" size={24} color="#fff" style={styles.buttonIcon} />
          <Text style={styles.loginButtonText}>Sign In</Text>
        </TouchableOpacity>

        <View style={styles.termsContainer}>
          <Text style={[styles.termsText, { color: currentColors.secondary }]}>
            By signing in, you agree to our Terms of Service and Privacy Policy
          </Text>
        </View>
      </View>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  content: {
    flex: 1,
    padding: 20,
    justifyContent: 'space-between',
    paddingVertical: 40,
  },
  logoContainer: {
    alignItems: 'center',
    marginTop: 40,
  },
  logo: {
    width: 550,
    height: 350,
    marginBottom: 20,
    alignItems: 'center'
  },
  appName: {
    fontSize: 32,
    fontWeight: 'bold',
    marginTop: 40,
  },
  tagline: {
    fontSize: 18,
    marginTop: 10,
    textAlign: 'center',
  },
  featuresContainer: {
    marginTop: 20,
    paddingHorizontal: 20,
  },
  featureItem: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 20,
    backgroundColor: 'rgba(255, 255, 255, 0.1)',
    padding: 15,
    borderRadius: 12,
  },
  featureText: {
    fontSize: 18,
    marginLeft: 15,
    flex: 1,
  },
  loginButton: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    padding: 16,
    borderRadius: 12,
    marginBottom: 20,
    marginHorizontal: 20,
  },
  buttonIcon: {
    marginRight: 10,
  },
  loginButtonText: {
    color: '#fff',
    fontSize: 18,
    fontWeight: '600',
  },
  termsContainer: {
    marginHorizontal: 20,
  },
  termsText: {
    textAlign: 'center',
    fontSize: 14,
  },
});