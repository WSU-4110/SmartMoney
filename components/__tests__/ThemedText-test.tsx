// components/__tests__/ThemedText-test.tsx
import React from 'react';
import renderer from 'react-test-renderer';
import { ThemedText } from '../ThemedText';
import { ThemeProvider } from '../../app/themeContext';
import { DefaultTheme } from '@react-navigation/native';

test('renders correctly', () => {
  const tree = renderer
    .create(
      <ThemeProvider>
        <ThemedText>Snapshot test!</ThemedText>
      </ThemeProvider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

