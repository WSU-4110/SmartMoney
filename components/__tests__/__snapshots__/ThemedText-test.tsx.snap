// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Text
  style={
    [
      {
        "color": "#120b07",
      },
      {
        "fontSize": 16,
        "lineHeight": 24,
      },
      undefined,
      undefined,
      undefined,
      undefined,
      undefined,
    ]
  }
>
  Snapshot test!
</Text>
`;
